hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/om/OMRatisLogRepair.java

```diff
@@ -0,0 +1,300 @@
+          }
+          if (oldRequest.hasVersion()) {
+            newRequest.setVersion(oldRequest.getVersion());
+          }
+
+          RaftProtos.StateMachineLogEntryProto oldEntry = next.getStateMachineLogEntry();
+          RaftProtos.StateMachineLogEntryProto.Builder newEntry =
+              RaftProtos.StateMachineLogEntryProto.newBuilder()
+                  .setCallId(oldEntry.getCallId())
+                  .setClientId(oldEntry.getClientId())
+                  .setType(oldEntry.getType())
+                  .setLogData(OMRatisHelper.convertRequestToByteString(newRequest.build()));
+          if (oldEntry.hasStateMachineEntry()) {
+            newEntry.setStateMachineEntry(oldEntry.getStateMachineEntry());
+          }
+
+          RaftProtos.LogEntryProto newLogEntry = RaftProtos.LogEntryProto.newBuilder()
+              .setTerm(next.getTerm())
+              .setIndex(next.getIndex())
+              .setStateMachineLogEntry(newEntry)
+              .build();
+
+          if (!isDryRun()) {
+            outputStream.write(newLogEntry);
+            outputStream.flush();
+          }
+          info("Replaced {" + oldRequest + "} with EchoRPC command at index "
+              + next.getIndex());
+        }
+      }
+
+      if (!isDryRun()) {
+        outputStream.flush();
+        outputStream.close();
+        Files.move(outputFile.toPath(), segmentFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+        info("Moved temporary output file to correct raft log location : " + segmentFile.toPath());
+      }
+
+    } catch (Exception ex) {
+      error("Exception: " + ex);
+    } finally {
+      if (logInputStream != null) {
+        logInputStream.close();
+      }
+      if (outputStream != null) {
+        outputStream.flush();
+        outputStream.close();
+      }
+      if (isDryRun()) {
+        boolean success = outputFile.delete();
+        if (!success) {
+          error("Error: Could not delete temporary output file \"" + outputFile + "\".");
+        }
+      }
+    }
+  }
+
+  private File createOutputFile(String name) throws IOException {
+    File temp = new File(name);
+    try {
+      if (temp.exists()) {
+        error("Warning: Temporary output file already exists - " + temp.getAbsolutePath() +
+            ". Trying to delete it and create a new one.");
+        boolean success = temp.delete();
+        if (!success) {
+          throw new IOException("Unable to delete old temporary file.");
+        }
+      }
+      boolean success = temp.createNewFile();
+      if (success) {
+        info("Temporary output file created successfully: " + temp.getAbsolutePath());
+      } else {
+        throw new IOException("createNewFile() failed.");
+      }
+    } catch (Exception e) {
+      throw new IOException("Error: Failed to create temporary output file - " + temp.getAbsolutePath(), e);
+    }
+    return temp;
+  }
+
+  private SegmentedRaftLogInputStream getInputStream(LogSegmentPath pi) {
```

Comment:
Can we use the public [LogSegment#readSegmentFile](https://github.com/apache/ratis/blob/688bbaee7b7b388c0e1eaf49b210a6a3877e4bd4/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java#L170) instead of this? We would provide a consumer of `LogEntryProto` to read the entries to process.